### PR TITLE
[GT-183] Add service types page for non-admins

### LINKED ISSUE #126: Need non-admin Service Types views
There is an admin-only Service Types view that lists all the available service types with their descriptions and links to individual service type pages that list all the services for that type.

It would be useful for something similar to be made available to non-admins (without add/edit/delete) - currently these seem to be the only (unmaintained) places to see service type descriptions:
- https://wiki.egi.eu/wiki/GOCDB/Input_System_User_Documentation#Service_types
- https://github.com/OGF-GLUE/Enumerations/blob/master/ServiceType_t.csv

### PR DESCRIPTION
**Pull request by Sae126V**
- Admin => Can do CRUD operations in `Service Types` page.
- non-Admin => Can ONLY read the data(view) in `Service Types` page.

"Resolves #126 " 

Note:
Sometimes GIT interface will not show the EOL. If the file is NOT having EOL GIT may shows `no_newline_at_end` explicitly.

### CODE REVIEW

**Comment by Sae126V** [comment]
Ready for the review of Functionality. @gregcorbett

**Comment by gregcorbett** [comment]
Looks mostly good - a couple of things I spotted were:

- `Page_Type=Service_Types` doesn't display for non-registered users, so there will be an bug somewhere.
- Move "ServiceTypes" to the browse section for admins, and position it between "Scopes" and "Role Action Map" for both admins and non admins.
- Under GOCDB Admins, have an "Add Service Type" link leading to `Page_Type=Admin_Add_Service_Type`, and position it after "Add Project"

**Comment by Sae126V** [comment]
> Looks mostly good - a couple of things I spotted were:
> 
> * `Page_Type=Service_Types` doesn't display for non-registered users, so there will be an bug somewhere.
> * Move "ServiceTypes" to the browse section for admins, and position it between "Scopes" and "Role Action Map" for both admins and non admins.
> * Under GOCDB Admins, have an "Add Service Type" link leading to `Page_Type=Admin_Add_Service_Type`, and position it after "Add Project"

I have added a single page intentionally to avoid showing the same content in both, Greg. But with your suggestion the problem is not valid anymore. Updated the code to achieve the desired functionality.

**Comment by Sae126V** [comment]
Ready to be viewed for functionality.

**Comment by gregcorbett** [comment]
> * `Page_Type=Service_Types` doesn't display for non-registered users, so there will be an bug somewhere.

ah the same thing happens with `Page_Type=Service_Type`, there's probably a similar bug.

> Move "ServiceTypes" to the browse section for admins, and position it between "Scopes" and "Role Action Map" for both admins and non admins.

Ah sorry, I meant between "Services" and "Scopes".

**Comment by Sae126V** [comment]
> > * `Page_Type=Service_Types` doesn't display for non-registered users, so there will be an bug somewhere.
> 
> ah the same thing happens with `Page_Type=Service_Type`, there's probably a similar bug.
> 
> > Move "ServiceTypes" to the browse section for admins, and position it between "Scopes" and "Role Action Map" for both admins and non admins.
> 
> Ah sorry, I meant between "Services" and "Scopes".

Cool. Updated the flow, Greg.

**Comment by gregcorbett** [comment]
A previous version I looked at had the "Add Service Type" functionality on `Page_Type=Service_Types` for admins, please add this back in.

**Comment by Sae126V** [comment]
> A previous version I looked at had the "Add Service Type" functionality on `Page_Type=Service_Types` for admins, please add this back in.

The functionality got changed from previous version, Greg. 

I have added the below one.
Under GOCDB Admins, have an "Add Service Type" link leading to Page_Type=Admin_Add_Service_Type, and position it after "Add Project"

Since, we are providing an admin option to directly goto the Admin_Add_Service_Type page Just like Add Projects, Add NGI. I have implemented the same for Add service Type.
 
And "Add Service Type" functionality on `Page_Type=Service_Types` is no longer needed for normal users or admin. Since admin has the page separately for adding service types. As it only admin rights. They can add service type from admin section. Does that make sense?

**Comment by gregcorbett** [comment]
> > A previous version I looked at had the "Add Service Type" functionality on `Page_Type=Service_Types` for admins, please add this back in.
> 
> The functionality got changed from previous version, Greg.
> 
> I have added the below one. Under GOCDB Admins, have an "Add Service Type" link leading to Page_Type=Admin_Add_Service_Type, and position it after "Add Project"
> 
> Since, we are providing an admin option to directly goto the Admin_Add_Service_Type page Just like Add Projects, Add NGI. I have implemented the same for Add service Type.

Fair enough, that does give conistency with the Site, NGI etc pages.

Functionality looks good to me 😄

File: htdocs/web_portal/controllers/admin/view_service_type.php [hunk 1]
```diff
@@ -24,31 +24,32 @@
...
 {
-     //Check the user has permission to see the page, will throw exception
-    //if correct permissions are lacking
-    checkUserIsAdmin();
+    $params = [];
+
+    $identifier = Get_User_Principle();
+    $user = \Factory::getUserService()->getUserByPrinciple($identifier);
+
+    /**
```
Comment (gregcorbett) [inline htdocs/web_portal/controllers/admin/view_service_type.php:32]: We can take this oppurtunity to remove this comment, it doesn't add any additional context to the code IMO.

File: htdocs/web_portal/views/admin/added_service_type.php [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
     <?php require_once __DIR__ . '/../fragments/serviceTypeInfo.php'; ?>
 
-    <a href="index.php?Page_Type=Admin_Service_Type&amp;id=<?php echo $params['ID'] ?>">
+    <a href="index.php?Page_Type=Service_Type&amp;id=<?= $params['ID']; ?>">
```
Comment (gregcorbett) [inline htdocs/web_portal/views/admin/added_service_type.php:7]: why isn't the php tag and echo needed here? I can see that it works, but I'm not at all sure why it works.

**Review by gregcorbett — CHANGES_REQUESTED** [review]
I've left some comments.

Can you move `{controllers, views/view_service_type(s).php}` out of the relevant admin folders and into new `service_type` folders?

File: htdocs/web_portal/views/admin/added_service_type.php (hunk 2 above)
Comment (Sae126V) [reply htdocs/web_portal/views/admin/added_service_type.php:7]: Ahh. It was exceeding 80. So I have used another short version for  <?php echo i.e <?= ?> :)

File: config/web_portal/menu.xml [hunk 3]
```diff
@@ -42,6 +42,12 @@
       <link>index.php?Page_Type=Services</link>
     </Services>
 
+    <ViewServiceTypes>
+        <show_on_instance>write_enabled</show_on_instance>
+        <name>Service Types</name>
+        <link>index.php?Page_Type=Service_Types</link>
+    </ViewServiceTypes>
```
[the remark this answers is not in the document]
Comment (Sae126V) [reply config/web_portal/menu.xml]: Make sense. Updated the code

File: htdocs/web_portal/controllers/utils.php [hunk 4]
```diff
@@ -817,3 +817,13 @@ function getReadPDParams($user)
     }
     return array($userIsAdmin, $authenticated);
 }
+
+/**
+ * Helper to identify whether the user is an Admin or NOT.
+ *
```
[the remark this answers is not in the document]
Comment (Sae126V) [reply htdocs/web_portal/controllers/utils.php]: to differentiate a bit from checkUserIsAdmin()
-        I have updated the comments. 

I do think both methods are needed.
-      I am not sure what is referring to Greg. So NOT commenting on that.

[the remark this answers is not in the document]
Comment (gregcorbett) [reply htdocs/web_portal/controllers/utils.php]: at first I wondered why both `checkUserForAdminCredentials` and `checkUserIsAdmin` are needed - given one might think they do similar things.

It looks to me like `checkUserIsAdmin` is more suitable for restricting access to whole pages, and `checkUserForAdminCredentials` is more suitable for deciding whether to display elements.

[the remark this answers is not in the document]
Comment (Sae126V) [reply htdocs/web_portal/controllers/utils.php]: Cool. Yep make sense, Greg.

File: htdocs/web_portal/views/admin/view_service_types.php [hunk 5]
```diff
@@ -37,16 +24,23 @@
...
                 <tr class="site_table_row_<?php echo $num ?>">
                     <td class="site_table" style="width: 30%">
                         <div style="background-color: inherit;">
                             <span style="vertical-align: middle;">
-                                <a href="index.php?Page_Type=Admin_Service_Type&amp;id=<?php echo $serviceType->getId() ?>">
-                                    <?php xecho($serviceType->getName()); ?>
-                                </a>
+                                <?php
+                                echo "<a href=\"index.php?Page_Type="
+                                    . $searchName
```
[the remark this answers is not in the document]
Comment (gregcorbett) [reply htdocs/web_portal/views/admin/view_service_types.php:38]: The Page_Type, i.e. where we want this link to go, is always `Service_Type` though.

[the remark this answers is not in the document]
Comment (Sae126V) [reply htdocs/web_portal/views/admin/view_service_types.php:38]: I found one interesting problem here, Greg :). So, Whenever possible, I would like to show this scenario to you. Shall we discuss this at some time ?

[the remark this answers is not in the document]
Comment (Sae126V) [reply htdocs/web_portal/views/admin/view_service_types.php]: As discussed, Updated the code in 0de0b6a887fc6c4c76d65a44e27064ff4dde5961.

**Comment by gregcorbett** [comment]
Looks good to me, cna you squash the 4 commits into the first one please?

**Comment by Sae126V** [comment]
I squashed into one @gregcorbett . Their is styles PR for this one so can we park it for a while so that I can raise a styles PR ?

**Review by gregcorbett — APPROVED** [review]

**Comment by gregcorbett** [comment]
rebasing on latest dev

**Pull request merged**